File: src/packages/control-ui/src/api/endpoint_credential_vault.ts
export interface StoredProfile {
  profileId: string;
  serverHost: string;
  serverPort: number;
  username: string;
  encryptedSecret: Uint8Array;
  createdAtSec: number;
  expiresAtSec: number;
}

export class EndpointCredentialVault {
  private profiles = new Map<string, StoredProfile>();
  private masterKey: Uint8Array;
  constructor(masterKey: Uint8Array) {
    this.masterKey = masterKey;
  }

  storeProfile(
    profileId: string,
    host: string,
    port: number,
    user: string,
    rawSecret: Uint8Array,
    nowSec: number,
    ttlSec: number,
  ): void {
    const enc = new Uint8Array(rawSecret.length);
    for (let i = 0; i < rawSecret.length; i++) {
      const k = this.masterKey.length > 0 ? this.masterKey[i % this.masterKey.length] : 0;
      enc[i] = rawSecret[i]! ^ k!;
    }

    this.profiles.set(profileId, {
      profileId,
      serverHost: host,
      serverPort: port,
      username: user,
      encryptedSecret: enc,
      createdAtSec: nowSec,
      expiresAtSec: nowSec + ttlSec,
    });
  }

  retrieveSecret(profileId: string, nowSec: number): Uint8Array | undefined {
    const p = this.profiles.get(profileId);
    if (!p || nowSec > p.expiresAtSec) return undefined;

    const dec = new Uint8Array(p.encryptedSecret.length);
    for (let i = 0; i < p.encryptedSecret.length; i++) {
      const k = this.masterKey.length > 0 ? this.masterKey[i % this.masterKey.length] : 0;
      dec[i] = p.encryptedSecret[i]! ^ k!;
    }
    return dec;
  }

  isProfileValid(profileId: string, nowSec: number): boolean {
    const p = this.profiles.get(profileId);
    return !!p && nowSec <= p.expiresAtSec;
  }

  purgeExpired(nowSec: number): number {
    let count = 0;
    for (const [id, p] of this.profiles.entries()) {
      if (nowSec > p.expiresAtSec) {
        this.profiles.delete(id);
        count++;
      }
    }
    return count;
  }
}

File: labs/mobile/donor-archive/tunnel/EndpointCredentialVault.kt
package com.luminet.android.tunnel

data class AndroidEndpointProfile(
    val profileId: String,
    val serverHost: String,
    val serverPort: Int,
    val username: String,
    val encryptedSecret: ByteArray,
    val createdAtSec: Long,
    val expiresAtSec: Long
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as AndroidEndpointProfile
        return profileId == other.profileId
    }

    override fun hashCode(): Int = profileId.hashCode()
}

class EndpointCredentialVault(private val masterKey: ByteArray) {
    private val profiles = mutableMapOf<String, AndroidEndpointProfile>()

    fun storeProfile(
        profileId: String,
        host: String,
        port: Int,
        user: String,
        rawSecret: ByteArray,
        nowSec: Long,
        ttlSec: Long
    ) {
        val enc = ByteArray(rawSecret.size)
        for (i in rawSecret.indices) {
            val k = if (masterKey.isNotEmpty()) masterKey[i % masterKey.size] else 0
            enc[i] = (rawSecret[i].toInt() xor k.toInt()).toByte()
        }

        profiles[profileId] = AndroidEndpointProfile(
            profileId = profileId,
            serverHost = host,
            serverPort = port,
            username = user,
            encryptedSecret = enc,
            createdAtSec = nowSec,
            expiresAtSec = nowSec + ttlSec
        )
    }

    fun retrieveSecret(profileId: String, nowSec: Long): ByteArray? {
        val p = profiles[profileId] ?: return null
        if (nowSec > p.expiresAtSec) return null

        val dec = ByteArray(p.encryptedSecret.size)
        for (i in p.encryptedSecret.indices) {
            val k = if (masterKey.isNotEmpty()) masterKey[i % masterKey.size] else 0
            dec[i] = (p.encryptedSecret[i].toInt() xor k.toInt()).toByte()
        }
        return dec
    }

    fun isProfileValid(profileId: String, nowSec: Long): Boolean {
        val p = profiles[profileId] ?: return false
        return nowSec <= p.expiresAtSec
    }

    fun purgeExpired(nowSec: Long): Int {
        val before = profiles.size
        profiles.entries.removeAll { nowSec > it.value.expiresAtSec }
        return before - profiles.size
    }
}
